include/openenclave/internal/defs.h

```diff
@@ -7,8 +7,22 @@
 #include <openenclave/bits/defs.h>
 
 /* OE_WEAK_ALIAS */
+#ifdef __GNUC__
 #define OE_WEAK_ALIAS(OLD, NEW) \
     extern __typeof(OLD) NEW __attribute__((__weak__, alias(#OLD)))
+#elif _MSC_VER
+#define OE_WEAK_ALIAS(OLD, NEW) \
+    __pragma(comment(linker, "/alternatename:" #NEW "=" #OLD))
```

Comment:
I cannot find any documentation about `alternatename`. So I am a bit hesitant about its use on the host side.
On the enclave side, this PR looks great!

Do we even need a weak directive?
Given two libraries `a.lib` and `b.lib` both containing a function `foo`, would the linker pick up the definition from the first library in the command line or would it complain of a multiple definition error?

Reply:
The linker will complain about multiple definitions if this is not included.

In theory you could declare stubs in a library and require that library to be linked in order to opt out, but that felt a bit backwards.

Reply:
My understanding is that the user opts in to switchless rather than opting out.
Is that not the case?

Reply:
I tested out MSVC, the linker prefers the definition in the first library, and does not emit a multiple definition error. This is consistent with the Linux linker behavior. Note: OE support for pluggable allocators is based on this linker behavior.

From my understanding of this PR,
- the empty stubs live in oecore
- the actual definitions live in oe_switchless.a/.lib that the enclave author has to link against.
Do we need a separate library just for the stubs?

Reply:
It is the case. I'm saying if you define a stub and then also import it in edl you will get a multiple definition error which is why weak symbols are necessary.

Reply:
> This is consistent with the Linux linker behavior

That's not the behavior I'm seeing. Where do pluggable allocators define a function twice? In pluggable allocators there is only ever one definition of the function, you just chose which one to use at link time, right?

> From my understanding of this PR,
>    the empty stubs live in oecore
>   the actual definitions live in oe_switchless.a/.lib that the enclave author has to link against.

This only works because the oecore version is weak. If I make both symbols strong I get linker errors.

Reply:
Interesting. It is not clear why an edl import would change the linker behavior in the above case.

Reply:
@mwshan Did the Intel SDK have to rely on weak symbols? From my understanding, Intel SDK supports opting out of "system edls".

Reply:
> Interesting. It is not clear why an edl import would change the linker behavior in the above case.

I don't understand. Can you point me to a place where a function is defined twice and doesn't cause a linker error? It is my understanding this will always cause a linker failure (at least in clang/gcc)

Reply:
> Interesting. It is not clear why an edl import would change the linker behavior in the above case.
> 
> I don't understand. Can you point me to a place where a function is defined twice and doesn't cause a linker error? It is my understanding this will always cause a linker failure (at least in clang/gcc)

1) The stubs are defined in `oecore` library. The actual implementations are in `oeswitchless` library. There are two definitions; one is intended to be replaced by the other. The linker picks the one in `oeswitchless`.

2) Similarly, `oecore` has implementations of allocator functions `oe_allocator_malloc`, `oe_allocator_free` etc which is implemented using `dlmalloc`. The library `oesnmalloc` implements the same functions, but using snmalloc. The linker picks up the definitions in `oesnmalloc` providing the ability to customize allocators.

3) Here is the GCC's published way to replace `malloc`. https://www.gnu.org/software/libc/manual/html_node/Replacing-malloc.html
There are multiple definitions of `malloc` - one provided by glibc and the one written by the user. The linker picks the one provided by the user, replacing GCC's malloc.

2) and 3) seem quite similar to 1). I am trying to understand what the difference is.

Reply:
3 works because globe defines these functions as weak symbols. Search for weak_alias here: https://github.com/lattera/glibc/blob/master/malloc/malloc.c

2 works because oedlmalloc and oesnmalloc aren’t both linked at the same time. It’s always one or the other.

I don’t believe 1 works without weak symbols.

Reply:
> This is consistent with the Linux linker behavior
> 
> That's not the behavior I'm seeing. Where do pluggable allocators define a function twice? In pluggable allocators there is only ever one definition of the function, you just chose which one to use at link time, right?
> 

There are **two** definitions, in two different libraries. One of them is chosen at link time.

> From my understanding of this PR,
> the empty stubs live in oecore
> the actual definitions live in oe_switchless.a/.lib that the enclave author has to link against.
> 
> This only works because the oecore version is weak. If I make both symbols strong I get linker errors.

Are the stub functions in a separate file? One reason I can think of that might cause a multiple definition error is if the stubs are placed in a file containing other functions that the linker has pulled in.

Reply:
> Are the stub functions in a separate file? One reason I can think of that might cause a multiple definition error is if the stubs are placed in a file containing other functions that the linker has pulled in.

After some more investigation this does seem to be what is happening.

But since the edl generated functions are defined as weak (to avoid conflicting with each other in the event that stubs for two enclaves are linked into the same host), the stub defined in oehost must also be weak. If it is strong and the others are weak, it will replace them.